File: src/types/assessment.ts
export type ProcessingStatus = 
  | "uploading"
  | "reading_question_paper"
  | "extracting_questions"
  | "reading_answers"
  | "detecting_regions"
  | "mapping_answers"
  | "preparing_assessment"
  | "completed"
  | "failed";

export interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface Question {
  id: string;
  number: string;
  text: string;
  page: number;
  bbox?: BoundingBox;
  confidence: number;
  sub_part?: string;
}

export interface AnswerRegion {
  page: number;
  bbox: BoundingBox;
}

export interface Answer {
  id: string;
  label: string;
  text?: string;
  pages: number[];
  regions: AnswerRegion[];
  confidence: number;
}

export type MappingMethod = 
  | "explicit_label"
  | "ocr_derived"
  | "structural"
  | "semantic"
  | "ai_reasoning";

export interface AnswerMapping {
  question_id: string;
  answer_id?: string;
  confidence: number;
  mapping_method?: MappingMethod;
}

export type QuestionStatus = "answered" | "unanswered" | "needs_review" | "partially_correct" | "incorrect" | "unable_to_determine";
export type AnswerStatus = "mapped" | "unmatched" | "needs_review";

export interface QuestionWithStatus {
  question: Question;
  status: QuestionStatus;
  mapping?: AnswerMapping;
  answer?: Answer;
  grading_info?: {
    status: QuestionStatus;
    score: number;
    satisfied_count: number;
    total_requirements: number;
    requirements: Array<{
      type: string;
      description: string;
      detected: boolean;
      item?: string;
    }>;
    satisfaction_results: Array<{
      satisfied: boolean;
      evidence: string;
    }>;
    recommendations: string[];
    explanation: string;
    evidence: string;
  };
}

export interface AssessmentResult {
  id: string;
  status: ProcessingStatus;
  questions: QuestionWithStatus[];
  unmatched_answers: Answer[];
  total_pages: number;
  processing_time_seconds?: number;
  error?: string;
  completeness_analysis?: {
    total_questions: number;
    answered: number;
    unanswered: number;
    needs_review: number;
    unable_to_determine: number;
    completion_rate: number;
    assessment: string;
  };
}

export interface UploadedFile {
  file: File;
  name: string;
  size: number;
  type: string;
  preview?: string;
}

File: backend/app/schemas/assessment.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any, Union
from enum import Enum


class ProcessingStatus(str, Enum):
    UPLOADING = "uploading"
    READING_QUESTION_PAPER = "reading_question_paper"
    EXTRACTING_QUESTIONS = "extracting_questions"
    READING_ANSWERS = "reading_answers"
    DETECTING_REGIONS = "detecting_regions"
    MAPPING_ANSWERS = "mapping_answers"
    PREPARING_ASSESSMENT = "preparing_assessment"
    COMPLETED = "completed"
    FAILED = "failed"


class BoundingBox(BaseModel):
    x: float = Field(..., ge=0, le=1, description="Normalized x coordinate (0-1)")
    y: float = Field(..., ge=0, le=1, description="Normalized y coordinate (0-1)")
    width: float = Field(..., ge=0, le=1, description="Normalized width (0-1)")
    height: float = Field(..., ge=0, le=1, description="Normalized height (0-1)")


class Question(BaseModel):
    id: str
    number: str
    text: str
    page: int
    bbox: Optional[BoundingBox] = None
    confidence: float = Field(..., ge=0, le=1)
    sub_part: Optional[str] = None


class AnswerRegion(BaseModel):
    page: int
    bbox: BoundingBox


class Answer(BaseModel):
    id: str
    label: str
    text: Optional[str] = None
    pages: List[int]
    regions: List[AnswerRegion]
    confidence: float = Field(..., ge=0, le=1)


class MappingMethod(str, Enum):
    EXPLICIT_LABEL = "explicit_label"
    OCR_DERIVED = "ocr_derived"
    STRUCTURAL = "structural"
    SEMANTIC = "semantic"
    AI_REASONING = "ai_reasoning"


class AnswerMapping(BaseModel):
    question_id: str
    answer_id: Optional[str] = None
    confidence: float = Field(..., ge=0, le=1)
    mapping_method: Optional[MappingMethod] = None


class QuestionStatus(str, Enum):
    ANSWERED = "answered"
    UNANSWERED = "unanswered"
    NEEDS_REVIEW = "needs_review"
    PARTIALLY_CORRECT = "partially_correct"
    INCORRECT = "incorrect"
    UNABLE_TO_DETERMINE = "unable_to_determine"


class AnswerStatus(str, Enum):
    MAPPED = "mapped"
    UNMATCHED = "unmatched"
    NEEDS_REVIEW = "needs_review"


class RequirementInfo(BaseModel):
    type: str
    description: str
    detected: bool
    item: Optional[str] = None

class SatisfactionResult(BaseModel):
    satisfied: bool
    evidence: str

class GradingInfo(BaseModel):
    status: QuestionStatus
    score: float
    satisfied_count: int
    total_requirements: int
    requirements: List[RequirementInfo]
    satisfaction_results: List[SatisfactionResult]
    recommendations: List[str]
    explanation: str
    evidence: str
    presentation_analysis: Optional[Dict[str, Any]] = None

class QuestionWithStatus(BaseModel):
    question: Question
    status: QuestionStatus
    mapping: Optional[AnswerMapping] = None
    answer: Optional[Answer] = None
    grading_info: Optional[GradingInfo] = None


class CompletenessAnalysis(BaseModel):
    total_questions: int
    answered: int
    unanswered: int
    needs_review: int
    unable_to_determine: int
    completion_rate: float
    assessment: str

class AssessmentResult(BaseModel):
    id: str
    status: ProcessingStatus
    questions: List[QuestionWithStatus]
    unmatched_answers: List[Answer]
    total_pages: int
    processing_time_seconds: Optional[float] = None
    error: Optional[str] = None
    completeness_analysis: Optional[CompletenessAnalysis] = None
